Flashing patched boot.img fails with ```(bootloader) Preflash validation failed```

so as the title says, flashing from fastboot fails with  ```(bootloader) Preflash validation failed```
but a different way of patching works for some reason, that is installing the latest 7 months old stable 27.0 build, patching and flashing boot.img with that build, updating the app to latest canary, do a direct install from magisk installation methods and the boot.img will be patched and flashed fine, i haven't uploaded recovery logs since i used fastboot but i have uploaded the patched boot.img that returns the error: https://file.io/EGEkNhXlsQJo

Device:Moto E7 Plus
Android version:14
Magisk version name:27.0
Magisk version code:27007

I've `git bisect`ed the problem down to between

https://github.com/topjohnwu/Magisk/compare/v27.0..06b6fb0c3

but I'm stuck now. 06b6fb0c3 isn't building for me. It's complaining about missing functions

<details>

```
ld.lld: error: undefined symbol: strerror
>>> referenced by magiskinit.773d76ed8ee366ad-cgu.0
>>>               lto.tmp:(base::xwrap::error_str::ha92e6618ee3abb9b (.llvm.4742393364852382803))

ld.lld: error: undefined symbol: __cxa_atexit
>>> referenced by mount.cpp:0 (src/init/mount.cpp:0)
>>>               lto.tmp:(_GLOBAL__sub_I_mount.cpp)
>>> referenced by rootdir.cpp:0 (src/init/rootdir.cpp:0)
>>>               lto.tmp:(_GLOBAL__sub_I_rootdir.cpp)
>>> referenced by rootdir.cpp:0 (src/init/rootdir.cpp:0)
>>>               lto.tmp:(_GLOBAL__sub_I_rootdir.cpp)
>>> referenced 2 more times

ld.lld: error: undefined symbol: pthread_setspecific
>>> referenced by magiskinit.773d76ed8ee366ad-cgu.0
>>>               lto.tmp:(std::sys_common::thread_local_key::StaticKey::set::h1280ce842977a72f)

ld.lld: error: undefined symbol: pthread_key_delete
>>> referenced by magiskinit.773d76ed8ee366ad-cgu.0
>>>               lto.tmp:(std::sys_common::thread_local_key::StaticKey::key::h0bc2f22e176e94df)
>>> referenced by magiskinit.773d76ed8ee366ad-cgu.0
>>>               lto.tmp:(std::sys_common::thread_local_key::StaticKey::key::h0bc2f22e176e94df)

ld.lld: error: undefined symbol: pthread_key_create
>>> referenced by magiskinit.773d76ed8ee366ad-cgu.0
>>>               lto.tmp:(std::sys::unix::thread_local_key::create::h854d4a62fbac876a)

ld.lld: error: undefined symbol: pthread_getspecific
>>> referenced by magiskinit.773d76ed8ee366ad-cgu.0
>>>               lto.tmp:(std::sys_common::thread_local_key::StaticKey::get::h15994ae9f8f1bc00)

ld.lld: error: undefined symbol: strerror_r
>>> referenced by magiskinit.773d76ed8ee366ad-cgu.0
>>>               lto.tmp:(_$LT$std..io..error..Error$u20$as$u20$core..fmt..Display$GT$::fmt::hc4ad587bcb3dfb7d (.llvm.4742393364852382803))

ld.lld: error: undefined symbol: pthread_cond_wait
>>> referenced by __threading_support:337 (src/external/libcxx/include/__threading_support:337)
>>>               lto.tmp:(all_xperm())

ld.lld: error: undefined symbol: pthread_cond_broadcast
>>> referenced by __threading_support:332 (src/external/libcxx/include/__threading_support:332)
>>>               lto.tmp:(all_xperm())

ld.lld: error: undefined symbol: pthread_mutex_lock
>>> referenced by __threading_support:306 (src/external/libcxx/include/__threading_support:306)
>>>               lto.tmp:(__cxxabiv1::(anonymous namespace)::InitByteGlobalMutex<__cxxabiv1::(anonymous namespace)::LibcppMutex, __cxxabiv1::(anonymous namespace)::LibcppCondVar, __cxxabiv1::(anonymous namespace)::GlobalStatic<__cxxabiv1::(anonymous namespace)::LibcppMutex>::instance, __cxxabiv1::(anonymous namespace)::GlobalStatic<__cxxabiv1::(anonymous namespace)::LibcppCondVar>::instance, &__cxxabiv1::(anonymous namespace)::PlatformThreadID()>::LockGuard::LockGuard(char const*))

ld.lld: error: undefined symbol: pthread_mutex_unlock
>>> referenced by __threading_support:316 (src/external/libcxx/include/__threading_support:316)
>>>               lto.tmp:(__cxxabiv1::(anonymous namespace)::InitByteGlobalMutex<__cxxabiv1::(anonymous namespace)::LibcppMutex, __cxxabiv1::(anonymous namespace)::LibcppCondVar, __cxxabiv1::(anonymous namespace)::GlobalStatic<__cxxabiv1::(anonymous namespace)::LibcppMutex>::instance, __cxxabiv1::(anonymous namespace)::GlobalStatic<__cxxabiv1::(anonymous namespace)::LibcppCondVar>::instance, &__cxxabiv1::(anonymous namespace)::PlatformThreadID()>::LockGuard::~LockGuard())

ld.lld: error: undefined symbol: android_set_abort_message
>>> referenced by abort_message.cpp:64 (src/external/libcxx/src/abi/abort_message.cpp:64)
>>>               lto.tmp:(abort_message)

ld.lld: error: undefined symbol: openlog
>>> referenced by abort_message.cpp:67 (src/external/libcxx/src/abi/abort_message.cpp:67)
>>>               lto.tmp:(abort_message)

ld.lld: error: undefined symbol: syslog
>>> referenced by abort_message.cpp:68 (src/external/libcxx/src/abi/abort_message.cpp:68)
>>>               lto.tmp:(abort_message)

ld.lld: error: undefined symbol: closelog
>>> referenced by abort_message.cpp:69 (src/external/libcxx/src/abi/abort_message.cpp:69)
>>>               lto.tmp:(abort_message)
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [/opt/android-sdk/ndk/ondk-r26.4/build/core/build-binary.mk:667: obj/init/local/armeabi-v7a/magiskinit] Error 1

```

</details>


( maybe because it's missing 0dbaf52566c5cac361ba60f12f1e05fea77bc824 ?? )

and if I try to avoid building armeabi-v7a with

```
--- a/build.py
+++ b/build.py
@@ -68,10 +68,10 @@ if shutil.which("ccache") is not None:
 cpu_count = multiprocessing.cpu_count()
 os_name = platform.system().lower()
 
-archs = ["armeabi-v7a", "x86", "arm64-v8a", "x86_64"]
+archs = ["arm64-v8a", "x86_64"]
 triples = [
-    "armv7a-linux-androideabi",
-    "i686-linux-android",
+    # "armv7a-linux-androideabi",
+    # "i686-linux-android",
     "aarch64-linux-android",
     "x86_64-linux-android",
 ]
```

then instead I get

> Android NDK: Module magisk depends on undefined modules: magisk-rs

--- a/native/src/boot/bootimg.cpp
+++ b/native/src/boot/bootimg.cpp
@@ -1019,9 +1019,35 @@ void repack(Utf8CStr src_img, Utf8CStr out_img, bool skip_comp) {
         memcpy(footer, boot.avb_footer, sizeof(AvbFooter));
         footer->original_image_size = __builtin_bswap64(aosp_img_size);
         footer->vbmeta_offset = __builtin_bswap64(off.vbmeta);
+
+        auto vbmeta = reinterpret_cast<AvbVBMetaImageHeader*>(out.data() + off.vbmeta);
+
         if (check_env("PATCHVBMETAFLAG")) {
-            auto vbmeta = reinterpret_cast<AvbVBMetaImageHeader*>(out.data() + off.vbmeta);
             vbmeta->flags = __builtin_bswap32(3);
+        }
+
+        // Sync hash descriptor image_size with the new AOSP portion size.
+        // Without this, some bootloaders (e.g. Motorola) reject images.
+        for (auto &desc : vbmeta->descriptors()) {
+            if (__builtin_bswap64(desc.tag) != AVB_DESCRIPTOR_TAG_HASH)
+                continue;
+
+            // enforce size limits; protect against adversarial input.
+            size_t buf_remaining = out.data() + out.size() - reinterpret_cast<uint8_t *>(&desc);
+            if (buf_remaining < __builtin_bswap64(desc.num_bytes_following) || buf_remaining - __builtin_bswap64(desc.num_bytes_following) < sizeof(AvbDescriptor)) {
+                // beware: both conditions are necessary because underflow in the subtraction could wrap
+                fprintf(stderr, "AVB hash descriptor num_bytes_following overflows buffer\n");
+                break;
+            }
+
+            if (__builtin_bswap64(desc.num_bytes_following) < sizeof(AvbHashDescriptor) - sizeof(AvbDescriptor)) {
+                fprintf(stderr, "AvbDescriptor too small to hold AvbHashDescriptor\n");
+                break;
+            }
+
+            auto &hd = reinterpret_cast<AvbHashDescriptor &>(desc);
+            hd.image_size = __builtin_bswap64(aosp_img_size);
+            break;
         }
     }
 

--- a/native/src/boot/bootimg.hpp
+++ b/native/src/boot/bootimg.hpp
@@ -69,6 +69,34 @@ struct AvbFooter {
     uint8_t reserved[28];
 } __attribute__((packed));
 
+// https://android.googlesource.com/platform/external/avb/+/refs/heads/android11-release/libavb/avb_descriptor.h
+enum AvbDescriptorTag : uint64_t {
+    AVB_DESCRIPTOR_TAG_PROPERTY        = 0,
+    AVB_DESCRIPTOR_TAG_HASHTREE        = 1,
+    AVB_DESCRIPTOR_TAG_HASH            = 2,
+    AVB_DESCRIPTOR_TAG_KERNEL_CMDLINE  = 3,
+    AVB_DESCRIPTOR_TAG_CHAIN_PARTITION = 4,
+};
+
+struct AvbDescriptor {
+    uint64_t tag;
+    uint64_t num_bytes_following;  // size of descriptor body (excludes this header); should always be a multiple of 8.
+} __attribute__((packed));
+
+// https://android.googlesource.com/platform/external/avb/+/refs/heads/android11-release/libavb/avb_hash_descriptor.h
+// for AvbDescriptor.tag == AVB_DESCRIPTOR_TAG_HASH
+struct AvbHashDescriptor {
+    AvbDescriptor header;
+    uint64_t image_size;
+    uint8_t hash_algorithm[32];
+    uint32_t partition_name_len;
+    uint32_t salt_len;
+    uint32_t digest_len;
+    uint32_t flags;
+    uint8_t reserved[60];
+    // followed by: partition_name, salt, digest (variable length)
+} __attribute__((packed));
+
 // https://android.googlesource.com/platform/external/avb/+/refs/heads/android11-release/libavb/avb_vbmeta_image.h
 struct AvbVBMetaImageHeader {
     uint8_t magic[AVB_MAGIC_LEN];
@@ -92,7 +120,47 @@ struct AvbVBMetaImageHeader {
     uint32_t rollback_index_location;
     uint8_t release_string[AVB_RELEASE_STRING_SIZE];
     uint8_t reserved[80];
+
+    struct AvbDescriptorRange descriptors();
 } __attribute__((packed));
+
+struct AvbDescriptorIterator {
+    AvbDescriptor *ptr;
+    AvbDescriptor &operator*() const { return *ptr; }
+    AvbDescriptor *operator->() const { return ptr; }
+    bool operator!=(const AvbDescriptorIterator &o) const {
+        if(ptr == nullptr) return false;
+        return ptr != o.ptr;
+    }
+    AvbDescriptorIterator &operator++() {
+        if (ptr->num_bytes_following % 8 != 0) {
+            // This is an error, a malformed image.
+            // https://android.googlesource.com/platform/external/avb/+/refs/heads/android11-release/libavb/avb_descriptor.h#60
+            // > For padding, |num_bytes_following| is always a multiple of 8.
+            // we can't signal an error easily, but we can stop the iteration.
+            ptr = nullptr;
+            return *this;
+        }
+        ptr = reinterpret_cast<AvbDescriptor *>(
+            reinterpret_cast<uint8_t *>(ptr) + sizeof(AvbDescriptor) + __builtin_bswap64(ptr->num_bytes_following));
+        return *this;
+    }
+};
+
+struct AvbDescriptorRange {
+    AvbDescriptor *first, *last;
+    AvbDescriptorIterator begin() const { return {first}; }
+    AvbDescriptorIterator end()   const { return {last}; }
+};
+
+inline AvbDescriptorRange AvbVBMetaImageHeader::descriptors() {
+    auto *base = reinterpret_cast<const uint8_t *>(this) + sizeof(AvbVBMetaImageHeader);
+    base += __builtin_bswap64(authentication_data_block_size);
+    base += __builtin_bswap64(descriptors_offset);
+    auto *first = reinterpret_cast<AvbDescriptor *>(const_cast<uint8_t *>(base));
+    auto *last  = reinterpret_cast<AvbDescriptor *>(const_cast<uint8_t *>(base) + __builtin_bswap64(descriptors_size));
+    return {first, last};
+}
 
 /*********************
  * Boot Image Headers

